designs/2018-processors-improvements/README.md

```diff
@@ -0,0 +1,271 @@
+
+### Processor Plugin Changes
+
+The first step is to change the format of processors in plugins. Currently, processors are exported by specifying the file extension to which the process should be applied. With this design, processors are exported by name, just like rules. For example:
+
+```js
+module.exports = {
+    processors: {
+        markdown: {
+            preprocess() {},
+            postprocess() {}
+        }
+    }
+};
+```
+
+This plugin now exports a named `markdown` processor that can be referenced directly from configuration. Existing plugins would need to be changed, but can be done in a backwards-compatible way (see Backwards Compatibility section below).
+
+### Add a `processors` Configuration Option
+
+In order to use a named processor, end-users would specify a `processors` option in their configuration. The option must be an array and must contain at least one string identifier. The string identifier must correspond to a named processor from a plugin and have the form *pluginName/processorName*, which would reference the `processorName` processor in the `eslint-plugin-pluginName` plugin. For example: 
+
+```js
+module.exports = {
+    plugins: ["markdown"],
+    parserOptions: {
+        ecmaVersion: 2018
+    },
+    processors: ["markdown/markdown"]
+};
+```
+
+This configuration file will apply the named processor `markdown` from the `eslint-plugin-markdown` plugin to all files.
+
+The `processors` option would be interpreted inside of `CLIEngine` to determine which processors to apply to files before linting should begin.
+
+### Apply Processors Using `overrides`
+
+Of course, in most cases the end-user will not want to apply a processor to all files. In that case, the `overrides` configuration option may be used to narrowly define to which files the processors should be applied. For example:
+
+```js
+module.exports = {
+    plugins: ["markdown"],
+    parserOptions: {
+        ecmaVersion: 2018
+    },
+    overrides: [
+        {
+            files: ["*.md"],
+            processors: ["markdown/markdown"]
+        }
+    ]
+};
+```
+
+In this configuration, the named processor `markdown` from the `eslint-plugin-markdown` plugin is applied only to files with the extension `*.md`. This mimics the behavior of the extension-named processor that `eslint-plugin-markdown` currently exports.
+
+The `CLIEngine` method `getConfigForFile()` would automatically get the correct value for `processors` when the configuration is calculated.
+
+**Note:** End-users would still need to provide `--ext .md` on the command line for ESLint to automatically lint files ending with `.md`. No part of this proposal changes the behavior of `--ext`.
+
+### Apply Multiple Processors
+
+In special cases, end-users may want to apply multiple processors to files. This can be accomplished by listing multiple named processors using the `processors` configuration option. For example:
```

Comment:
Oh yes, you are correct. So I guess what we really need is a `processor` configuration key that is just a string because the rest of the logic will be determined by matching `overrides`.